### PR TITLE
Issue #2317 - Aligned representation of activatable dependencies with ActivePropertyValidator

### LINKED ISSUE #2317: Reference hierarchy: support for displaying active referencing entities only
### Description

The Reference Hierarchy (original issue #1426) provides a way to see and traverse all entities that reference a given entity instance. Some of the referencing entities may be activatable with many instances being inactive. The current implementation does not distinguish between active and inactive references.

- [x] 1. The main objective of this issue is to enhance the functionality of the Reference Hierarchy to display only active referencing entities that represent dependencies, preventing deactivation of a referenced entity. The resultant entities, presented by the Reference Hierarchy should be fully aligned with those identified by`ActivePropertyValidator`.

    image

- [x] 2. Provide support for the Reference Hierarchy to be invoked as a property action from Entity Masters.

### Change overview

`ReferenceHierarchyWebUiConfig` has two new action factory methods to create property action configurations:
1. `mkPropAction` — intended to be used if a property action is required for a entity-typed property; such action would produce a reference hierarchy for an entity, entered in that property; if associated with a non-entity-typed property, the reference hierarchy assumes the master entity (i.e., `this`) as the referred entity.
2. `mkPropActionForMasterEntity` — intended to be used to create a property action that pertains to the master entity (i.e., `this`); in other words, regardless of the actual property type such action is configured for, the reference hierarchy would be invoked for `this`; a special provision exists for property `active` of activatable entities – a reference hierarchy action would have property `activeOnly` (i.e., "Active dependencies only?") defaulted to `true`.

Please note that the semantics for `activeOnly` is such that only active activatable references that prevent entity deactivation are considered, filtering out any inactive activatables, activatables that do not impact deactivation (e.g., due to `@SkipActivatableTracking`), and non-activatable entities. This logic is fully aligned with `ActivePropertyValidator` as adjusted in issue #2316.

image

### Expected outcome
Ability to show only active activatable dependencies, preventing entity deactivation, or all referencing entities.

### PR DESCRIPTION
**Pull request by 01es**
Completely resolve #2317.

# To be completed by the pull request creator

This section should be completed with reference to section [Preparing PR](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews#preparing-pr) of the [Code and PR reviews](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews) wiki page.

- [x] Converted PR to draft while it is being prepared by tapping the "Convert to draft" link beneath the "Reviewers" section.

- [x] A self-review of all changes has been completed, and the changes are in sync with the issue requirements.

- [x] Changes to the requirements have been reflected in the issue description.

- [x] Any "leftovers" such as sysouts, printing of stack traces, and any other "temporary" code, have been removed.

- [x] Minor refactorings, such as renamings, extraction of constants, etc., have been addressed.

- [x] Developer documentation (e.g., comments, Javadoc), have been provided where required.

- [x] All tests (all existing tests and all new tests) pass successfully.

- [x] Changes have been inspected for possible NPE situations, and the changes are sufficiently defensive.

- [x] The correct base branch has been selected for these changes to be merged into.

- [x] The latest changes from the base branch have already been merged into this feature branch (and tested).

- [x] Added a change overview to the issue description or as a wiki page, referenced in the issue description.
      Some issue might be very descriptive and server in place of a wiki page.
      In such cases consider adding label `Wiki like` to the issue.

- [x] Changes subject to performance considerations have been evaluated, and tested against production-size data if applicable.

- [x] This pull request does not contain significant changes, and at least one appropriate reviewer has been selected.

- [x] The `In progress` label has been removed from the issue.

- [x] The `Pull request` label has been added to the issue.

- [x] Made PR ready for review by tapping the "Ready for review" button below the list of commits on the PR page.

# To be completed by the pull request reviewer

This section should be completed with reference to section [Performing PR review](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews#performing-pr-review) of the [Code and PR reviews](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews) wiki page.

- [x] The `In progress` label has been added to the pull request in GitHub.

- [x] The issue requirements have been read and understood (along with any relevant emails and/or Slack messages).

- [x] The correct base branch is specified, and that base branch is up-to-date in the local source.

- [x] The issue branch has been checked out locally, and had the base branch merged into it.

- [x] All automated tests pass successfully.

- [x] Ensure the implementation satisfies the functional requirements.

- [x] Ensure that code changes are secure and align with the established coding practices, including code formatting and naming conventions.

- [x] Ensure that code changes are documented and covered with automated tests as applicable.

- [x] Ensure that code changes are well-suited for informal reasoning.

- [x] Ensure that changes are documented for the end-user (a software engineer in the case of TG, or an application user in the case of TG-based applications).

- [x] If there are significant changes (described above), special attention has been paid to them.
      Marked the task items in section "Significant changes" as completed to indicated that corresponding changes have been reviewed, improved if necessary, and approved.

- [x] The issue or issues addressed by the pull request are associated with the relevant release milestone.

# To be completed by the pull request reviewer once the changes have been reviewed and accepted

- [ ] The changes have been merged into the base branch (unless there is a specific request not to do so, e.g., they are to be released to SIT).

- [ ] The issue branch has been deleted (unless the changes have not been merged - see above, or there is a specific request not to do so).

- [x] The `In progress` label has been removed from the pull request.

- [ ] The `Pull request` label has been removed from the pull request.

### CODE REVIEW

File: platform-dao/src/main/java/ua/com/fielden/platform/ref_hierarchy/ReferenceHierarchyDao.java [hunk 1]
```diff
@@ -79,18 +81,49 @@ public ReferenceHierarchy save(final ReferenceHierarchy action) {
...
+                                                              DomainEntityDependency::propPath,
+                                                              toSet()
+                                                          )));
+                // Populate dependencies based on the inverted dependencies.
+                // For example, for Person, we get:
+                // WorkActivity -> {Person -> [referredTo, manager.person, returnedToServiceBy.person, technician.person, authoriser]}, which includes the relevant property paths.
+                invertedDependencies.forEach((referencingEntityType, properties) ->
+                                    {
+                                        dependencies.put(referencingEntityType, Map.of(referencedEntityType, properties));
+                                    });
```
Comment (elvin-fms) [inline platform-dao/src/main/java/ua/com/fielden/platform/ref_hierarchy/ReferenceHierarchyDao.java:110]: Are the curly braces aka brackets really necessary here?

File: platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/query/metadata/DataDependencyQueriesGenerator.java [hunk 2]
```diff
@@ -54,8 +54,11 @@ public static Optional<QueryExecutionModel<EntityAggregates, AggregatedResultQue
             return empty();
         }
 
-        final AggregatedResultQueryModel qry = select(queries).groupBy().prop("type").yield().prop("type").as("type").yield().countAll().as("qty").modelAsAggregate();
-        return Optional.of(from(qry).with(orderBy().yield("qty").desc().model()).model());
+        final AggregatedResultQueryModel qry = select(queries).groupBy().prop("type").groupBy().prop("simpleType")
```
Comment (elvin-fms) [inline platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/query/metadata/DataDependencyQueriesGenerator.java:57]: Perhaps each `.groupBy()` also deserves to be on a separate line?

**Review by elvin-fms — COMMENTED** [review]
1. Should the "References" branch even appear in the "Active dependencies only?" mode. The entities, which appear there, do not appear to satisfy the "Active dependencies" criteria.

2. In the change overview, the reference to "Show active only?" should probably be replaced with "Active dependencies only?"

3. Are unit tests required to cover the new/changed functionality?

File: platform-dao/src/main/java/ua/com/fielden/platform/ref_hierarchy/ReferenceHierarchyDao.java (hunk 1 above)
Comment (01es) [reply platform-dao/src/main/java/ua/com/fielden/platform/ref_hierarchy/ReferenceHierarchyDao.java:110]: It was for ease of my personal comprehension in this case to highlight the fact of mutation.

File: platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/query/metadata/DataDependencyQueriesGenerator.java (hunk 2 above)
Comment (01es) [reply platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/query/metadata/DataDependencyQueriesGenerator.java:57]: If you think that it does -- don't hesitate to make it so.

**Comment by 01es** [comment]
> 1. Should the "References" branch even appear in the "Active dependencies only?" mode. The entities, which appear there, do not appear to satisfy the "Active dependencies" criteria.

Issue clearly states that the new option should affect "Referenced By" only. Ergo, yes – the "References" branch remains there according to issue requirements. If you would like for "References" to be removed if the option if activated, please create a separate issue.

> 2. In the change overview, the reference to "Show active only?" should probably be replaced with "Active dependencies only?"

Certainly. And was that difficult for you to make that tiny improvement? Especially that we were discussing the functional changes to this issue together with you.

> 3. Are unit tests required to cover the new/changed functionality?

Regrettably the Reference Hierarchy does not have a separate suit of unit tests. A separate technical debt issue should be created for this, if it does not exist already. @oleh-maikovych, please check and capture such issue if required.

Lickily, the functionality of the Reference Hierarchy had extensive interactive testing.

**Review by oleh-maikovych — APPROVED** [review]

**Pull request merged**